rust/libs/logging/src/windows_event_log.rs

```diff
@@ -0,0 +1,456 @@
+
+        let Ok(guard) = self.handle.lock() else {
+            return;
+        };
+        let Some(handle) = *guard else {
+            return;
+        };
+
+        // SAFETY: Handle is valid and protected by mutex.
+        unsafe {
+            let _ = ReportEventW(
+                handle,
+                event_type,
+                0,
+                event_id,
+                None,
+                0,
+                Some(&messages),
+                None,
+            );
+        }
+    }
+}
+
+impl Drop for EventSourceHandle {
+    fn drop(&mut self) {
+        if let Some(handle) = self.handle.get_mut().expect("not poisoned").take() {
+            // SAFETY: We own the handle.
+            unsafe {
+                let _ = DeregisterEventSource(handle);
+            }
+        }
+    }
+}
+
+/// A tracing layer that writes events to the Windows Event Log.
+pub struct WindowsEventLogLayer {
+    handle: Arc<EventSourceHandle>,
+}
+
+impl WindowsEventLogLayer {
+    fn new(source: &str) -> Result<Self, Error> {
+        let _ = try_register_source(source);
+
+        let source_wide = to_wide_string(source);
+        let handle = unsafe { RegisterEventSourceW(PCWSTR::null(), PCWSTR(source_wide.as_ptr())) }
+            .map_err(|e| Error::OpenSource {
+                source: source.to_owned(),
+                error: e,
+            })?;
+
+        Ok(Self {
+            handle: Arc::new(EventSourceHandle::new(handle)),
+        })
+    }
+}
+
+/// Storage for span fields.
+#[derive(Default)]
+struct SpanFields {
+    fields: Vec<(String, String)>,
+}
+
+impl<S> Layer<S> for WindowsEventLogLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let mut fields = SpanFields::default();
+            attrs.record(&mut FieldVisitor(&mut fields.fields));
+            span.extensions_mut().insert(fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &tracing::span::Record<'_>, ctx: Context<'_, S>) {
+        if let Some(span) = ctx.span(id) {
+            let mut extensions = span.extensions_mut();
+            if let Some(fields) = extensions.get_mut::<SpanFields>() {
+                values.record(&mut FieldVisitor(&mut fields.fields));
+            }
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        let level = event.metadata().level();
+
+        let (event_type, event_id) = match *level {
+            Level::ERROR => (EVENTLOG_ERROR_TYPE, EVENT_ID_ERROR),
+            Level::WARN => (EVENTLOG_WARNING_TYPE, EVENT_ID_WARNING),
+            _ => (EVENTLOG_INFORMATION_TYPE, EVENT_ID_INFO),
+        };
+
+        let mut message = None;
+        let mut fields = Vec::new();
+        event.record(&mut EventVisitor {
+            message: &mut message,
+            fields: &mut fields,
+        });
```

Comment:
I'd implement `Default` for `EventVisitor` here and instantiate that and then pass a single mutable reference into `record`.